### ISSUE TITLE
net/http: possibly unaligned 64-bit word accessed atomically

### ISSUE THREAD

**Issue by pam4** [golang/go#33821]
[Here](https://github.com/golang/go/blob/f3e3b71a50d7a247d07cc4e4295ec2dc44b04b6a/src/net/http/server.go#L292) it is.

If I understand [the alignment problem](https://golang.org/pkg/sync/atomic/#pkg-note-BUG) correctly, a struct nested in another struct is not by itself "allocated", therefore you cannot assume its first word is 64-bit aligned.

However there is a fair chance that I'm missing something, see the [CL](https://go-review.googlesource.com/c/go/+/121419/). It seems it was done quite on purpose.

If I'm misunderstanding the alignment problem, please close this and I'll file a bug for the `atomic` documentation because the "allocated" wording in the `Bugs` section is very confusing.

Notice that `sync.WaitGroup` goes to all the trouble of making an aligned 64-bit word out of a `[3]uint32` using `unsafe.Pointer`, [here](https://github.com/golang/go/blob/4c3f26076b6a9853bcc3c7d7e43726c044ac028a/src/sync/waitgroup.go#L23).

Another suspicious 64-bit word accessed atomically is field [numClosed](https://github.com/golang/go/blob/dc63b59630cbc7fe1b039757bac0d6f713dfc26d/src/database/sql/sql.go#L411) in the `DB` struct of `database/sql`. I'm not sure it's safe to assume that its preceding field size is a multiple of 64 bit (interface type).

I'm concerned about other cases too. For example I'm not familiar with the `expvar` package, but if a user makes a struct with a field of type `expvar.Int` (not a pointer), he's going to cause trouble. Is it obvious that you can't do that?

**Comment by ianlancetaylor** [comment]
The code in net/http works because the field is comes after an even number of pointers.  Same for `DB.NumClosed`.  This code is certainly fragile and there ought to be a better way.  One proposal is #19057 but it hasn't gone anywhere.  There is also #599.

I'm going to close this issue since there isn't a bug here.

**Comment by pam4** [comment]
@ianlancetaylor, sorry I don't get it.

- In the [CL](https://go-review.googlesource.com/c/go/+/121419/1/src/net/http/server.go#286) you commented *"I don't see anything keeping this aligned to a 64-bit boundary on a 32-bit system"*, and the field was already at the same exact position.
Then @bradfitz changed it from type `uint64` to type `struct{ atomic uint64 }` and it was fixed. What's the meaning of that change?

- How do you count the even number of pointers?
Before `conn.curState` I see *5* pointers, *2* interfaces, *2* strings, *1* `func()`, and *1* struct from an external package.
Before `DB.numClosed` I see *1* `int64` and *1* interface.
There's not even a comment about this near the affected fields.

I'm aware of the issues you mentioned. I doubt there's something I haven't read already on the subject, but it's still a big confusion.
What about the `expvar` package I mentioned?

**Comment by ianlancetaylor** [comment]
I think that I wasn't thinking when looking at that CL.  I agree that I don't see why Brad's change helps the problem.

We have a general problem with atomic access to 64-bit words on 32-bit systems, and that needs a general solution.  I encourage you to open an issue about that, especially if it is not already covered by the issues I mentioned above.  I'm not sure it helps to open an issue about a specific problem in net/http.

**Comment by pam4** [comment]
@ianlancetaylor, thanks for the clarification.

> I'm not sure it helps to open an issue about a specific problem in net/http.

Well, while we are at it, I think moving `curState` on top of the `conn` struct, with a nice comment like `// keep this 64-bit aligned`, and removing the misleading nested struct would help a lot.

It seems far too likely that the next contributor changing `conn` layout will miscalculate or even overlook the whole problem.

I found that code while looking at the standard library as a reference, to consolidate my understanding of the problem, and I got misled by it instead. Other people might waste time for the same reason.
